Makefile

```diff
@@ -596,6 +635,17 @@ startup_files: include_dirs $(LIBC_BOTTOM_HALF_CRT_OBJS)
 	mkdir -p "$(SYSROOT_LIB)" && \
 	cp $(LIBC_BOTTOM_HALF_CRT_OBJS) "$(SYSROOT_LIB)"
 
+# TODO: As of this writing, wasi_thread_start.s uses non-position-independent
+# code, and I'm not sure how to make it position-independent.  Once we've done
+# that, we can enable libc.so for the wasi-threads build.
+ifneq ($(THREAD_MODEL), posix)
+LIBC_SO = \
+	$(SYSROOT_LIB)/libc.so \
+	$(SYSROOT_LIB)/libwasi-emulated.so
+endif
+
+libc_so: include_dirs $(LIBC_SO)
```

Comment:
is this the main target to build shared library?
which target builds pic version of crt1*.o?

Reply:
> is this the main target to build shared library? 
Yes

> which target builds pic version of crt1*.o?
My understanding is that crt1.o is intended to be linked into the main module, which doesn't necessarily need to be PIC since it's the one exporting memory, not importing it.  I'll admit I've been focusing on reactor-style apps, though, so if there's something missing to support command-style apps I'm happy to add it.

Reply:
> > which target builds pic version of crt1*.o?
> 
> My understanding is that crt1.o is intended to be linked into the main module, which doesn't necessarily need to be PIC since it's the one exporting memory, not importing it. I'll admit I've been focusing on reactor-style apps, though, so if there's something missing to support command-style apps I'm happy to add it.

i was assuming that `-pie` was used to produce a dynamically linked executable.
https://github.com/WebAssembly/tool-conventions/blob/main/DynamicLinking.md#llvm-implementation
but as you say, it might be possible to produce a non-pic dynamically linked executable.
i haven't tried.

btw, a reactor module is linked with crt1-reactor.o as well.
i guess you meant shared libraries, for which if crt1-reactor.o should be linked isn't obvious.

Reply:
> but as you say, it might be possible to produce a non-pie dynamically linked executable.
i haven't tried.

it seems clang can somehow produce non-pie dynamically linked main module.
eg.
```
-fPIC
-nodefaultlibs
-Xlinker --unresolved-symbols=import-dynamic
-Xlinker --export-table
-Xlinker --growable-table
```

a few things are not obvious to me though:
* what to do for `__stack_pointer`. should the main module export it? or import it?
* how about `__heap_base` and `__heap_end`?

on the other hand, pie main module is more straightforward and known to work as it's what emscripten `-sMAIN_MODULE` uses.
to produce pie main modules, we need pic (or at least `dynamic-no-pic`) version of crt1.

how do you think?

Reply:
Yes, I believe the main module must export `__stack_pointer`, `__heap_base`, and `__heap_end` (or at least _some_ module must export them, because shared libraries will import them in general).

I'd be fine with building crt1-*.o with `-fPIC` unconditionally -- it's a very small amount of code, so I doubt there will be a noticeable performance penalty.  Does anyone have concerns about that?

Reply:
As a data point, emscripten currently always builds the main module as `-pie` and then the runtime provides things like `__stack_pointer`, `__heap_end`, etc.  I've been trying for a while to convert emscripten to use statically linked main modules but I have not had any success yet.  See https://github.com/emscripten-core/emscripten/issues/12682

Reply:
FWIW, `wasm-tools component link` synthesizes a main module which exports `_stack_pointer` etc. based on a set of input shared libraries.  In other words, all the application code and its dependencies are built as shared libraries and "pseudo-dynamically" linked together into a component, so I don't believe crt1.o plays any role in the process.  I realize that's not the _only_ way WASI shared libraries might be used, but that's the use case I've been focusing on so far.

Reply:
I think `crt1.o` (or something like it) will always be needed in order to define the `_initialize` function (who primary job, at least in the llvm world is to call `__wasm_call_ctors`).

Reply:
> I think `crt1.o` (or something like it) will always be needed in order to define the `_initialize` function (who's primary job, at least in the llvm world is to call `__wasm_call_ctors`).

`wasm-tools component link` also synthesizes an "init" module which contains a `start` function that calls `__wasm_apply_data_relocs`, `__wasm_call_ctors`, initializes `GOT` globals, etc.

Reply:
I think in the long run we should use something like `_initialize` exported rather than `__wasm_call_ctors` (which I see as an llvm-specific thing).

`__wasm_apply_data_relocs` is a tricky one because (at least in cyclic module graphs) we need to call that on the whole module graph before we call `__wasm_call_ctors` on the whole module graph.. so perhaps we need to standardize on some kind of phase 1 and phase 2 `_initialize` functions.

Reply:
I've pushed an update to build `crt1-*.o` with `-fPIC`.